### PR TITLE
[2/N] Port 5 _composable distributed test to Intel GPU

### PR DESCRIPTION
**Pull request by zxd1997066**
For https://github.com/pytorch/pytorch/issues/114850, we will port distributed tests to Intel GPU. This is the second PR for _composable cases, the first is https://github.com/pytorch/pytorch/pull/159118.
We could enable Intel GPU with following methods and try the best to keep the original code styles:

- Use "torch.accelerator.current_accelerator()" to determine the accelerator backend
- Enabled XPU for some test path
- Skip some test cases which Intel GPU does not support
- Added "cpu:gloo,xpu:xccl" for distributed backend

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci @tianyu-l @XilunWu @gujinghui @EikanWang @fengyuan14 @guangyey

### CODE REVIEW

File: test/distributed/_composable/test_composability/test_2d_composability.py [hunk 1]
```diff
@@ -58,6 +58,11 @@
 from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir
 
 
+device = torch.accelerator.current_accelerator()
+device_module = torch.get_device_module(device)
```
Comment (guangyey) [inline test/distributed/_composable/test_composability/test_2d_composability.py]: Where is `device_module` used?

File: test/distributed/_composable/test_composability/test_2d_composability.py [hunk 2]
```diff
@@ -58,6 +58,11 @@
 from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir
 
 
+device = torch.accelerator.current_accelerator()
```
Comment (guangyey) [inline test/distributed/_composable/test_composability/test_2d_composability.py]:
```suggestion
device_type = acc.type if (acc := torch.accelerator.current_accelerator()) else "cpu"
```

File: test/distributed/_composable/test_composability/test_2d_composability.py (hunk 1 above)
Comment (zxd1997066) [reply test/distributed/_composable/test_composability/test_2d_composability.py]: > Where is `device_module` used?

will remove it.

File: test/distributed/_composable/test_replicate_with_compiler.py [hunk 3]
```diff
@@ -98,6 +98,8 @@ def _test_compile(
         self.create_pg(device)
         torch._dynamo.config.optimize_ddp = "python_reducer"
         torch.manual_seed(123)
+        if device_type == "xpu":
```
Comment (guangyey) [inline test/distributed/_composable/test_replicate_with_compiler.py:101]: If without this code change, which test case will fail due to the non-deterministic accuracy issue.

Comment (zxd1997066) [reply test/distributed/_composable/test_replicate_with_compiler.py:101]: test_compile_backward_only, test_compile_bf16, test_compile_fp16, test_compile_gpu, test_compile_gpu_ac, listed in https://github.com/intel/torch-xpu-ops/issues/1668

> If without this code change, which test case will fail due to the non-deterministic accuracy issue.

File: test/distributed/_composable/test_replicate.py [hunk 4]
```diff
@@ -273,14 +283,14 @@ def forward(self, x: torch.Tensor):
...
+        device_module.set_device(self.rank)
         dim = 3
         bz = 2
-        model = ToyModel(dim).cuda()
+        model = ToyModel(dim).to(device)
         for linear in model.linears:
             fully_shard(linear)
         fully_shard(model.linears)
-        replicate(model, device_id=torch.cuda.current_device())
+        replicate(model, device_id=device_module.current_device())
```
Comment (guangyey) [inline test/distributed/_composable/test_replicate.py]:
```suggestion
        replicate(model, device_id=torch.accelerator.current_device_index())
```

File: test/distributed/_composable/test_replicate.py [hunk 5]
```diff
@@ -273,14 +283,14 @@ def forward(self, x: torch.Tensor):
                 return y
 
         self._init_pg()
-        torch.cuda.set_device(self.rank)
+        device_module.set_device(self.rank)
```
Comment (guangyey) [inline test/distributed/_composable/test_replicate.py]:
```suggestion
        torch.accelerator.set_device_index(self.rank)
```

File: test/distributed/_composable/test_replicate.py [hunk 6]
```diff
@@ -233,13 +242,13 @@ def test_replicate_device_id(self):
...
 
-        replicate(model_cuda, device_id=torch.device(torch.cuda.current_device()))
+        replicate(model_cuda, device_id=torch.device(device_module.current_device()))
         # DDP instance is attached in first pre forward
         model_cuda(torch.randn(2, 2))
         replicate_ddp_weakref = replicate.state(model_cuda)._ddp_weakref()
         self.assertEqual([0], replicate_ddp_weakref.device_ids)
         # Pass in int as device_id
-        replicate(model_cuda2, device_id=int(torch.cuda.current_device()))
+        replicate(model_cuda2, device_id=int(device_module.current_device()))
```
Comment (guangyey) [inline test/distributed/_composable/test_replicate.py]: ditto

File: test/distributed/_composable/test_replicate.py [hunk 7]
```diff
@@ -233,13 +242,13 @@ def test_replicate_device_id(self):
         # Should be None for CPU training
         self.assertEqual(None, replicate_ddp_weakref.device_ids)
 
-        replicate(model_cuda, device_id=torch.device(torch.cuda.current_device()))
+        replicate(model_cuda, device_id=torch.device(device_module.current_device()))
```
Comment (guangyey) [inline test/distributed/_composable/test_replicate.py]: ditto

File: test/distributed/_composable/test_replicate.py [hunk 8]
```diff
@@ -166,24 +173,25 @@ def forward(self, inp, *, kwarg=None):
...
         # CPU input ensures replicate can move arg and kwargs to device.
         a, b = torch.randn(2, 2), torch.randn(2, 2)
         model(a, kwarg=b).sum().backward()
 
     @skip_if_lt_x_gpu(2)
+    @unittest.skipIf(TEST_XPU, "XPU does not support gloo backend")
     def test_replicate_ignore_module(self):
         self._init_pg()
-        torch.cuda.set_device(self.rank)
+        device_module.set_device(self.rank)
```
Comment (guangyey) [inline test/distributed/_composable/test_replicate.py]: ditto

File: test/distributed/_composable/test_replicate.py [hunk 9]
```diff
@@ -166,24 +173,25 @@ def forward(self, inp, *, kwarg=None):
                 return self.a(inp)
 
         self._init_pg()
-        torch.cuda.set_device(self.rank)
-        model = MyNet().cuda()
-        replicate(model, device_id=torch.cuda.current_device())
+        device_module.set_device(self.rank)
```
Comment (guangyey) [inline test/distributed/_composable/test_replicate.py]: ditto

File: test/distributed/_composable/test_replicate.py [hunk 10]
```diff
@@ -166,24 +173,25 @@ def forward(self, inp, *, kwarg=None):
...
+        model = MyNet().to(device)
+        replicate(model, device_id=device_module.current_device())
```
Comment (guangyey) [inline test/distributed/_composable/test_replicate.py]: ditto

**Review by guangyey — COMMENTED** [review]
Please use `torch.accelerator` API as much as possible.

**Comment by zxd1997066** [comment]
> Please use `torch.accelerator` API as much as possible.

Thanks, will modify it.

File: test/distributed/_composable/test_checkpoint.py [hunk 11]
```diff
@@ -22,16 +26,16 @@ def __init__(self, device: torch.device):
 
     def __enter__(self):
         self.active_memory_enter = (
-            torch.cuda.memory_stats()["active_bytes.all.current"]
-            if self.device.type == "cuda"
+            device_module.memory_stats()["active_bytes.all.current"]
```
Comment (guangyey) [inline test/distributed/_composable/test_checkpoint.py]:
```suggestion
            torch.accelerator.memory_stats()["active_bytes.all.current"]
```

File: test/distributed/_composable/test_checkpoint.py [hunk 12]
```diff
@@ -22,16 +26,16 @@ def __init__(self, device: torch.device):
...
+            if self.device.type == "cuda" or self.device.type == "xpu"
             else 0
         )
         return self
 
     def __exit__(self, *exc):
         self.active_memory_exit = (
-            torch.cuda.memory_stats()["active_bytes.all.current"]
-            if self.device.type == "cuda"
+            device_module.memory_stats()["active_bytes.all.current"]
```
Comment (guangyey) [inline test/distributed/_composable/test_checkpoint.py]:
```suggestion
            torch.accelerator.memory_stats()["active_bytes.all.current"]
```

**Review by guangyey — APPROVED** [review]
LGTM.

File: test/distributed/_composable/test_checkpoint.py [hunk 13]
```diff
@@ -10,10 +10,14 @@
 import torch.nn as nn
 from torch.distributed._composable import checkpoint
 from torch.testing._internal.common_cuda import TEST_CUDA
-from torch.testing._internal.common_utils import run_tests, TestCase
+from torch.testing._internal.common_utils import run_tests, TEST_XPU, TestCase
 from torch.utils.checkpoint import CheckpointError
 
 
+device_type = acc.type if (acc := torch.accelerator.current_accelerator()) else "cpu"
+device_module = torch.get_device_module(device_type)
```
Comment (guangyey) [inline test/distributed/_composable/test_checkpoint.py]: unused?

Comment (zxd1997066) [reply test/distributed/_composable/test_checkpoint.py]: yes, removed it.

**Review by d4l3k — APPROVED** [review]
LGTM
